### PR TITLE
fix: requestValidation when hostname URL has any uppercase letters

### PR DESCRIPTION
**Pull request by thinkingserious**
Fixes https://github.com/twilio/twilio-csharp/issues/522

### Checklist
- [x] I acknowledge that all my contributions will be made under the project's license

Internal Reference: DI-1369

### CODE REVIEW

File: Dockerfile.dev [hunk 1]
```diff
@@ -0,0 +1,28 @@
+FROM python:3.7.5-stretch
```
Comment (childish-sambino) [inline Dockerfile.dev:1]: Why is the existing Dockerfile not sufficient?

File: test/Twilio.Test/Security/RequestValidatorTest.cs [hunk 2]
```diff
@@ -31,6 +31,14 @@ public void TestValidateDictionary()
             Assert.IsTrue(_validator.Validate(Url, _parameters, signature), "Request does not match provided signature");
         }
 
+        [Test]
+        public void TestValidateDictionaryMixedCase()
+        {
+            const string signature = "RSOYDt4T1cUTdK1PDd93/VVr8B8=";
+            const string url = "https://MyCompany.com:1234/MyApp.PHP?foo=1&bar=2";
+            Assert.IsFalse(_validator.Validate(url, _parameters, signature), "Request should have failed validation but didn't");
```
Comment (eshanholtz) [inline test/Twilio.Test/Security/RequestValidatorTest.cs:39]: Don't we want to test the case where mixed case succeeds?

File: src/Twilio/Security/RequestValidator.cs [hunk 3]
```diff
@@ -129,17 +129,17 @@ private string RemovePort(UriBuilder uri)
...
         private string AddPort(UriBuilder uri)
         {
             if (uri.Port != -1)
             {
-                return uri.ToString();
+                return uri.Uri.OriginalString;
             }
             uri.Port = uri.Scheme == "https" ? 443 : 80;
-            return uri.ToString();
+            return uri.Uri.OriginalString;
```
Comment (eshanholtz) [inline src/Twilio/Security/RequestValidator.cs:142]: This returns the original string passed to the constructor, so wouldn't this end up ignoring the addition of the port?

File: Dockerfile.dev (hunk 1 above)
Comment (thinkingserious) [reply Dockerfile.dev:1]: I get this error:

```
Could not find 'mono' host. Make sure that 'mono' is installed on the machine and is available in PATH environment variable.

Test Run Aborted.
```

I do have `mono` version `6.12.0.122` on the host machine.

File: src/Twilio/Security/RequestValidator.cs (hunk 3 above)
Comment (thinkingserious) [reply src/Twilio/Security/RequestValidator.cs:142]: I'm not sure, will double check the results of the TestValidateAddsPortHttp/s tests.

Comment (thinkingserious) [reply src/Twilio/Security/RequestValidator.cs:154]: Refactored this based on these test results using [this tool](https://docs.microsoft.com/en-us/dotnet/api/system.uri.originalstring?view=net-5.0 ):

```csharp
Uri uriAddress = new Uri("https://mycompany.com/myapp.php?foo=1&bar=2");
UriBuilder uriBuilder = new UriBuilder(uriAddress);
Console.WriteLine(uriAddress.ToString());
Console.WriteLine(uriBuilder.Uri.OriginalString);
// Resulting string: https://mycompany.com/myapp.php?foo=1&bar=2
Match m = Regex.Match(uriAddress.OriginalString, "^(([^:/?#]+):)?(//([^/?#]*))?([^?#]*)(\\?([^#]*))?(#(.*))?");
var hostName = m.Groups[4].ToString();
uriBuilder.Host = hostName.Split(':')[0];
uriBuilder.Port = uriBuilder.Scheme == "https" ? 443 : 80;
Console.WriteLine(uriBuilder.Uri.OriginalString);
// Resulting string: https://mycompany.com:443/myapp.php?foo=1&bar=2
uriBuilder.Host = hostName.Split(':')[0];
uriBuilder.Port = -1;
Console.WriteLine(uriBuilder.Uri.OriginalString);
// Resulting string: https://mycompany.com/myapp.php?foo=1&bar=2
 
uriAddress = new Uri("https://MyCompany.com/myapp.php?foo=1&bar=2");
uriBuilder = new UriBuilder(uriAddress);
Console.WriteLine(uriAddress.ToString());
Console.WriteLine(uriBuilder.Uri.OriginalString);
// Resulting string: https://mycompany.com:443/myapp.php?foo=1&bar=2
m = Regex.Match(uriAddress.OriginalString, "^(([^:/?#]+):)?(//([^/?#]*))?([^?#]*)(\\?([^#]*))?(#(.*))?");
hostName = m.Groups[4].ToString();
uriBuilder.Host = hostName.Split(':')[0];
uriBuilder.Port = uriBuilder.Scheme == "https" ? 443 : 80;
Console.WriteLine(uriBuilder.Uri.OriginalString);
// Resulting string: https://MyCompany.com:443/myapp.php?foo=1&bar=2
uriBuilder.Host = hostName.Split(':')[0];
uriBuilder.Port = -1;
Console.WriteLine(uriBuilder.Uri.OriginalString);
// Resulting string: https://MyCompany.com/myapp.php?foo=1&bar=2
```

File: test/Twilio.Test/Security/RequestValidatorTest.cs (hunk 2 above)
Comment (thinkingserious) [reply test/Twilio.Test/Security/RequestValidatorTest.cs:80]: Added more tests.

File: Dockerfile.dev (hunk 1 above)
Comment (thinkingserious) [reply Dockerfile.dev:1]: Updated the `CONTRIBUTING.md` guidance with some additional gotchas and solutions. However, I could not get the existing `Dockerfile` to work locally, only the new `Dockerfile.dev`

File: src/Twilio/Security/RequestValidator.cs [hunk 4]
```diff
@@ -124,22 +125,33 @@ private static bool SecureCompare(string a, string b)
             return mismatch == 0;
         }
 
-        private string RemovePort(UriBuilder uri)
+        private string PreserveHostnameCasing(string url)
```
Comment (childish-sambino) [inline src/Twilio/Security/RequestValidator.cs:128]: Should move this to below `Add/RemovePort` so the code reads from top-to-bottom.

File: src/Twilio/Security/RequestValidator.cs [hunk 5]
```diff
@@ -124,23 +125,33 @@ private static bool SecureCompare(string a, string b)
             return mismatch == 0;
         }
 
-        private string RemovePort(UriBuilder uri)
+        private string RemovePort(string url)
         {
-            // UriBuilder.ToString() will not display the port 
-            // if the Port property is set to -1
+            var uri = new UriBuilder(url);
+            uri.Host = PreserveHostnameCasing(url);
```
Comment (childish-sambino) [inline src/Twilio/Security/RequestValidator.cs:131]: Why is this needed if `OriginalString` doesn't change the casing?

Comment (thinkingserious) [reply src/Twilio/Security/RequestValidator.cs:131]: Turns out that it still lower cases the host. Please see the comment above to Elise for details.

Comment (childish-sambino) [reply src/Twilio/Security/RequestValidator.cs:131]: What about other parts of the URL, like the scheme or credentials?

Comment (thinkingserious) [reply src/Twilio/Security/RequestValidator.cs:131]: Those are preserved, I'm only modifying the host. I will add some tests to verify.

Comment (thinkingserious) [reply src/Twilio/Security/RequestValidator.cs:131]: Added a test case that has credentials in the URL. Schemes were already being tested along with ports.

File: test/Twilio.Test/Security/RequestValidatorTest.cs (hunk 2 above)
Comment (thinkingserious) [reply test/Twilio.Test/Security/RequestValidatorTest.cs:88]: Generating the signatures using Python via:

```python
from twilio.request_validator import RequestValidator

validator = RequestValidator("12345")
uri = "http://username:password@MyCompany.com/myapp.php?foo=1&bar=2"
params = {
    "CallSid": "CA1234567890ABCDE",
    "Digits": "1234",
    "From": "+14158675309",
    "To": "+18005551212",
    "Caller": "+14158675309",
}
print(validator.compute_signature(uri, params, utf=False))
```

File: src/Twilio/Security/RequestValidator.cs (hunk 5 above)
Comment (childish-sambino) [reply src/Twilio/Security/RequestValidator.cs:131]: Don't see any tests with an uppercase scheme or creds.

Comment (thinkingserious) [reply src/Twilio/Security/RequestValidator.cs:131]: Ah, I see. 👀

Comment (thinkingserious) [reply src/Twilio/Security/RequestValidator.cs:131]: Just added a few more tests. One with upper cased scheme and mixed case credentials.

Comment (thinkingserious) [reply src/Twilio/Security/RequestValidator.cs:131]: Turns out the credentials were preserved, but the scheme was not. Also needed to add logic to account for the existence of credentials when trying to extract the host.

File: src/Twilio/Security/RequestValidator.cs [hunk 6]
```diff
@@ -124,23 +127,48 @@ private static bool SecureCompare(string a, string b)
...
             uri.Port = -1;
-            return uri.ToString();
+            var scheme = PreserveSchemeCasing(url);
+            return uri.Uri.OriginalString.Replace(uri.Scheme, scheme);
```
Comment (childish-sambino) [inline src/Twilio/Security/RequestValidator.cs:136]: Why is this doing a `Replace` instead of updating the `uri.Scheme`, like is done for `uri.host`?

File: src/Twilio/Security/RequestValidator.cs [hunk 7]
```diff
@@ -124,23 +127,48 @@ private static bool SecureCompare(string a, string b)
...
+            var hostName = m.Groups[4].ToString();
+            var preservedHostName = System.String.Empty;
+            // Do we have credentials in the URL?
+            if (hostName.Split('@').Length == 2)
+            {
+                preservedHostName = hostName.Split('@')[1].Split(':')[0];
+            } else {
+                preservedHostName = hostName.Split(':')[0];
+            }
+            return preservedHostName;
```
Comment (childish-sambino) [inline src/Twilio/Security/RequestValidator.cs:164]: Think rather than trying to regex parse and extract out pieces of the URL (which is what `UriBuilder` is also doing), the logic can be simplified by getting the substring from the original URL case insensitive. E.g., `uri.Host = url.substring(url.toLowerCase().indexOf(uri.Host), uri.Host.length());` Just in C# and not Java syntax.

File: src/Twilio/Security/RequestValidator.cs (hunk 6 above)
Comment (thinkingserious) [reply src/Twilio/Security/RequestValidator.cs:136]: `uri.Scheme` does not let us override like `uri.Host` :(

It returns a lower cased value.

File: src/Twilio/Security/RequestValidator.cs [hunk 8]
```diff
@@ -124,23 +124,45 @@ private static bool SecureCompare(string a, string b)
...
         }
 
+        // uri.Scheme will lowercase the output
+        private string PreserveSchemeCasing(string url)
+        {
+            var uri = new UriBuilder(url);
+            var startIndex = url.IndexOf(uri.Scheme, StringComparison.OrdinalIgnoreCase);
+            var preservedScheme = url.Substring(startIndex, uri.Scheme.Length);
+            return preservedScheme;
+        }
```
Comment (childish-sambino) [inline src/Twilio/Security/RequestValidator.cs:166]: This can be refactored to a single function which takes in the url and the part needing to be replaced: `PreserveCasing(url, uri.Host)` and `PreserveCasing(url, uri.Scheme)`.

File: src/Twilio/Security/RequestValidator.cs [hunk 9]
```diff
@@ -124,23 +124,45 @@ private static bool SecureCompare(string a, string b)
...
-            return uri.ToString();
+            return SetPort(url, -1);
         }
 
-        private string AddPort(UriBuilder uri)
+        private string AddPort(string url)
         {
-            if (uri.Port != -1)
+            var uri = new UriBuilder(url);
+            return SetPort(url, uri.Port);
```
[the remark this answers is not in the document]
Comment (thinkingserious) [reply src/Twilio/Security/RequestValidator.cs:135]: Updated the logic and the test `TestValidateDictionaryMixedCaseWithPort`.

File: src/Twilio/Security/RequestValidator.cs [hunk 10]
```diff
@@ -124,23 +124,37 @@ private static bool SecureCompare(string a, string b)
...
+
+        private string SetPort(string url, int port)
+        {
+            var uri = new UriBuilder(url);
+            uri.Host = PreserveCase(url, uri.Host);
+            if (port == -1)
             {
-                return uri.ToString();
+                uri.Port = port;
+            } else if ( port != 443 || port != 80 ) {
```
Comment (childish-sambino) [inline src/Twilio/Security/RequestValidator.cs:145]: This will always yield `true`.

File: test/Twilio.Test/Security/RequestValidatorTest.cs [hunk 11]
```diff
@@ -82,9 +155,9 @@ public void TestValidateRemovesPortHttps()
         public void TestValidateRemovesPortHttp()
         {
             const string url = "http://mycompany.com:1234/myapp.php?foo=1&bar=2";
-            Assert.IsTrue(_validator.Validate(url, _parameters, "Zmvh+3yNM1Phv2jhDCwEM3q5ebU="), "Request does not match provided signature");
+            Assert.IsTrue(_validator.Validate(url, _parameters, "OyGYPqTF6ztdbRNCvuQO/oPvqQ4="), "Request does not match provided signature");
```
[the remark this answers is not in the document]
Comment (childish-sambino) [reply test/Twilio.Test/Security/RequestValidatorTest.cs:158]: This was an existing test (regression). The signature should not have changed.

File: src/Twilio/Security/RequestValidator.cs [hunk 12]
```diff
@@ -124,23 +124,37 @@ private static bool SecureCompare(string a, string b)
...
+
+        private string SetPort(string url, int port)
+        {
+            var uri = new UriBuilder(url);
+            uri.Host = PreserveCase(url, uri.Host);
+            if (port == -1)
             {
-                return uri.ToString();
+                uri.Port = port;
+            } else if ( port != 443 && port != 80 ) { // we have a custom port
```
Comment (childish-sambino) [inline src/Twilio/Security/RequestValidator.cs:145]: nit: spacing and newlines inconsistent.

**Review by childish-sambino — APPROVED** [review]
